web/src/message_view.js

```diff
@@ -661,6 +661,86 @@ export function show(raw_terms, opts) {
     }
 }
 
+function handle_anchor_case(anchor, has_found_anchor, get_msg_id) {
+    // These functions are scoped inside `handle_anchor_case` to
+    // to avoid them being used for any other purpose.
+    function duplicate_current_msg_list_with_new_data(data) {
+        const msg_list = new message_list.MessageList({data});
+        msg_list.reading_prevented = message_lists.current.reading_prevented;
+        return msg_list;
+    }
+
+    function select_msg_id(msg_id, select_opts) {
+        message_lists.current.select_id(msg_id, {
+            then_scroll: true,
+            from_scroll: false,
+            ...select_opts,
+        });
+    }
+
+    function select_anchor_using_data(data) {
+        const msg_list = duplicate_current_msg_list_with_new_data(data);
+        message_lists.update_current_message_list(msg_list);
+        // `force_rerender` is required to render the new data.
+        select_msg_id(get_msg_id(), {force_rerender: true});
+    }
+
+    if (has_found_anchor(message_lists.current.data.fetch_status)) {
+        select_msg_id(get_msg_id());
+    } else {
+        let msg_list_data = new MessageListData({
+            filter: message_lists.current.data.filter,
+            excludes_muted_topics: message_lists.current.data.excludes_muted_topics,
+        });
+        load_local_messages(msg_list_data);
+
+        if (has_found_anchor(all_messages_data.fetch_status)) {
+            select_anchor_using_data(msg_list_data);
+        } else {
+            msg_list_data = new MessageListData({
+                filter: message_lists.current.data.filter,
+                excludes_muted_topics: message_lists.current.data.excludes_muted_topics,
+            });
+            // We are fetching either `FETCH_SIZE` messages forward or backward from the anchor
+            // for currently handled anchors (oldest/newest). It is important that we
+            // are quick to make the jump feel fast while keeping enough context.
+            const FETCH_SIZE = 50;
```

Comment:
I don't see why we would want to use this value rather than the standard constants in `message_fetch.ts` as though we were narrowing? Generally fetching a bit more history doesn't change the API response time much, and I think 50 has the property that we may end up fetching twice-in-a-row if the user scrolls a bit. Probably 100 would be a fine number, aka `narrowed_view_backward_batch_size`

In any case, let's define this number in `message_fetch.consts`, maybe by making a small helper in that file that this calls -- we really want to keep that configuration centralized.

Reply:
Makes sense. Done!